walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,90 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // throw new UnsupportedOperationException("TODO: Implement this method."); 
+    long reqDuration = request.getDuration();
+    Collection<String> reqAttendees = request.getAttendees();
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // Special cases
+    if (reqDuration > TimeRange.WHOLE_DAY.duration()) {
+        return possibleTimes;
+    }
+
+    // Collection of clashing events' (events where at least one attendee belongs to reqAttendees) timeranges
+    ArrayList<TimeRange> clashingEventTimes = new ArrayList<TimeRange>();
+
+    // Check each event in events collection
+    for (Event event : events) {
+        Set<String> eventAttendees = event.getAttendees();
+
+        // Check if any of the requested attendees need to attend this event
+        boolean isClashingEvent = false;
+        for (String attendee : eventAttendees) {
+            if (reqAttendees.contains(attendee)) {
+                isClashingEvent = true;
+                break;
+            }
+        }
+        if (isClashingEvent == false) continue;
+        clashingEventTimes.add(event.getWhen());
+    }
+
+    // Check for case where there is no clashing event
+    int numEvents = clashingEventTimes.size();
+    if (numEvents == 0) {
+        possibleTimes.add(TimeRange.WHOLE_DAY);
+        return possibleTimes;
+    }
+
+    // Sort by ascending start time
+    Collections.sort(clashingEventTimes, TimeRange.ORDER_BY_START);
+
+    // First check if there is space before the start of the first event
+    int startOfFirstEvent = clashingEventTimes.get(0).start();
+    if (startOfFirstEvent >= reqDuration) {
```

Comment:
Since you do this many times, consider adding a helper:

TimeRange getPossibleTimeRange(start, end) 

that returns null if the duration isn’t long enough

Reply:
Thanks for the suggestion, I directly added the new TimeRange into possibleTimes in the method so that I don't have to add checks for null TimeRanges.